### ISSUE TITLE
User support ticket system

### ISSUE THREAD

**Issue by brainwane** [pypa/warehouse#3231]
PyPI users should have a place within or connected to Warehouse to submit support tickets (account recovery, increasing upload file size limits, requesting new trove classifiers, etc.). Right now, we get support requests via IRC, GitHub issues, Twitter, Slack, SourceForge issues, and probably other places too. Support requests should be in their own central queue, separate from Warehouse bug reports and feature requests.

Ideally this tool would allow a user to choose whether to submit a report privately, e.g., to report a user breaching the terms of service, that would only be visible to PyPI admins, or to submit it publicly and then link to it from their home community.

From [today's Warehouse core developers meeting](https://wiki.python.org/psf/PackagingWG/2018-03-12-Warehouse) and connected to #3016 and probably #2017 and #1919.

**Comment by ncoghlan** [comment]
Looking at this question from a broader PSF perspective, it might also be valuable to talk to the folks that handle mail.python.org as to whether it might help them if `postmaster` and the various `*-owner` contact lists could opt-in to being hooked up to a ticketing system of some kind. (/cc @msapiro)

As things currently stand, I suspect @MarkMangoba and the other PSF staff don't have much (if any) visibility into that work, and hence not only can't identify opportunities where the PSF might be able to help streamline things, but also can't suitably acknowledge those volunteers for their efforts.

**Comment by brainwane** [comment]
@ewdurbin Is there current PSF infrastructure that could, without too much scope creep, handle this? I'd love to get this resolved to help with forward motion on https://github.com/pypa/warehouse/issues/1506#issuecomment-375736028 .

**Comment by ewdurbin** [comment]
I believe @di started a lightweight ticketing system built into warehouse at PyCon, but I'm unsure on status.

Options I see:

## Build it into warehouse itself

This has the advantage of letting us manage specific forms that help users provide administrators the right information to complete a given support concern, but is something more to maintain.

## Deploy a ticketing system to the PSF infrastructure

Would be helpful for other projects and support concerns that don't fit public issue trackers like GitHub well, but comes with the burden of upkeep and maintenance.

If there's a known solution that works well I'm all ears!

## Pay for or solicit donation of a Support "solution"

Generally more pleasant than a self-hosted system, but costly if we pay out of pocket (most run $10-15 per user per month).

I've reached out to a few companies without much luck. I can do another round of emailing to see if someone would support our cause!

**Comment by di** [comment]
Yep, I'm working on something that would allow users to flag a project or request an upload limit, and allow administrators to approve/deny. The advantage to this approach is that approving the request can just perform the requested action directly, rather than having to jump from a ticketing system to the PyPI admin UI.

It's about 1/2 done, and I'm hoping to finish it up in the upcoming weeks.

**Comment by pradyunsg** [comment]
> flag a project

Does flagging a project refer to name-squats or in general request transfers?

**Comment by di** [comment]
@pradyunsg It would simply be a way to raise an issue to the PyPI admins referencing a specific project, so it could be for name-squats, malicious or offensive packages, or just to request ownership of an abandoned project.

**Comment by ncoghlan** [comment]
Another instance of this at https://github.com/pypa/packaging-problems/issues/172

**Comment by ncoghlan** [comment]
One for an ownership transfer: https://github.com/pypa/warehouse/issues/5088

**Comment by brainwane** [comment]
> Yep, I'm working on something that would allow users to flag a project or request an upload limit, and allow administrators to approve/deny. The advantage to this approach is that approving the request can just perform the requested action directly, rather than having to jump from a ticketing system to the PyPI admin UI.
> 
> It's about 1/2 done, and I'm hoping to finish it up in the upcoming weeks.

@di could you share your branch? Thanks!

**Comment by di** [comment]
Here's the branch I was referring to... it's quite out of date at this point:

https://github.com/pypa/warehouse/compare/master...di:flag-projects

I attempted to start with just the ability to flag a project or request an upload limit, with the assumption that this could grow into an additional support system.

I believe I stopped at the point where I was trying to get the state machine correct, and also generalizable. Happy to dig back into this more if someone wants to pick up where I left off.

**Comment by brainwane** [comment]
Since so many other features are blocked on this, I [added it to the list of fundable packaging projects](https://wiki.python.org/psf/Fundable%20Packaging%20Improvements#User_support_ticket_system) on the wiki.

**Comment by brainwane** [comment]
@di @ewdurbin @yeraydiazdiaz @jamadden How does https://github.com/pypa/pypi-support fit into all this? Should we link to it in https://pypi.org/help/#twofa (right now it says "If you lose your authentication application and can no longer log in, the PyPI team cannot currently help you recover your account.")?
